File: typescript/ringbuf-ts/tests/ringbuf.test.ts
import { test } from 'node:test';
import * as assert from 'node:assert';
import * as path from 'path';
import * as os from 'os';
import * as fs from 'fs';
import * as net from 'net';
import {
    RingBuffer,
    Writer,
    Reader,
    newConnection,
    ROLE_HOST,
    ROLE_PLUGIN,
    TimeoutError,
    SocketSignalManager
} from '../src/index.js';

async function localSocketPair(): Promise<[net.Socket, net.Socket]> {
    const server = net.createServer();
    await new Promise<void>((resolve) => server.listen(0, '127.0.0.1', resolve));
    const port = (server.address() as net.AddressInfo).port;

    const acceptPromise = new Promise<net.Socket>((resolve) => {
        server.on('connection', (socket) => {
            server.close();
            resolve(socket);
        });
    });

    const client = net.connect(port, '127.0.0.1');
    const serverSocket = await acceptPromise;

    return [serverSocket, client];
}

function getTempFilePath(prefix: string): string {
    return path.join(os.tmpdir(), `${prefix}_${Math.random().toString(36).substring(2, 9)}`);
}

test('RingBuffer - Basic', async () => {
    const filePath = getTempFilePath('shm_basic');
    const rb = new RingBuffer(filePath, 2, 64);
    rb.clear();

    const [s1, s2] = await localSocketPair();
    const sigMgr1 = new SocketSignalManager(s1);
    const sigMgr2 = new SocketSignalManager(s2);

    try {
        const payload1 = Buffer.from('hello world');
        await rb.write(payload1, 42n, 10, sigMgr1);

        const [readPayload, seq] = await rb.read(10, sigMgr2);
        assert.strictEqual(seq, 42n);
        assert.deepStrictEqual(readPayload, payload1);
    } finally {
        sigMgr1.close();
        sigMgr2.close();
        rb.close();
        try { fs.unlinkSync(filePath); } catch (e) {}
    }
});

test('RingBuffer - Timeout and Drop', async () => {
    const filePath = getTempFilePath('shm_timeout');
    const rb = new RingBuffer(filePath, 2, 64);
    rb.clear();

    const [s1, s2] = await localSocketPair();
    const sigMgr1 = new SocketSignalManager(s1);
    const sigMgr2 = new SocketSignalManager(s2);

    try {
        await rb.write(Buffer.from('msg1'), 1n, 10, sigMgr1);
        await rb.write(Buffer.from('msg2'), 2n, 10, sigMgr1);

        // Buffer full, write expects timeout
        await assert.rejects(
            async () => {
                await rb.write(Buffer.from('msg3'), 3n, 20, sigMgr1);
            },
            (err) => err instanceof TimeoutError
        );

        // writeIndex should advance to 1
        assert.strictEqual(rb.writeIndex, 1);
    } finally {
        sigMgr1.close();
        sigMgr2.close();
        rb.close();
        try { fs.unlinkSync(filePath); } catch (e) {}
    }
});

test('Connection - Basic', { skip: process.platform === 'win32' }, async () => {
    const baseDir = getTempFilePath('shm_conn');
    fs.mkdirSync(baseDir, { recursive: true });
    const basePath = path.join(baseDir, 'shm_comm');

    let hostWriter: Writer | null = null;
    let hostReader: Reader | null = null;
    let pluginWriter: Writer | null = null;
    let pluginReader: Reader | null = null;

    try {
        const hostPromise = newConnection(basePath, ROLE_HOST, 2, 64, 500, 500);
        // Wait a tiny bit for server listener to bind
        await new Promise(r => setTimeout(r, 20));
        const pluginPromise = newConnection(basePath, ROLE_PLUGIN, 2, 64, 500, 500);

        [hostWriter, hostReader] = await hostPromise;
        [pluginWriter, pluginReader] = await pluginPromise;

        // Host writes, Plugin reads
        const msgFromHost = Buffer.from('hello from host');
        await hostWriter.write(msgFromHost);

        const readBufPlugin = await pluginReader.read(100);
        assert.deepStrictEqual(readBufPlugin, msgFromHost);

        // Plugin writes, Host reads
        const msgFromPlugin = Buffer.from('hello from plugin');
        await pluginWriter.write(msgFromPlugin);

        const readBufHost = await hostReader.read(100);
        assert.deepStrictEqual(readBufHost, msgFromPlugin);

    } finally {
        if (hostWriter) hostWriter.close();
        if (hostReader) hostReader.close();
        if (pluginWriter) pluginWriter.close();
        if (pluginReader) pluginReader.close();

        // Cleanup
        try {
            fs.rmSync(baseDir, { recursive: true, force: true });
        } catch (e) {}
    }
});

test('RingBuffer - Large Data Integrity', async () => {
    const filePath = getTempFilePath('shm_large');
    const rb = new RingBuffer(filePath, 2, 64);
    rb.clear();

    const [s1, s2] = await localSocketPair();
    const sigMgr1 = new SocketSignalManager(s1);
    const sigMgr2 = new SocketSignalManager(s2);

    const writer = new Writer(rb, sigMgr1, 1000);
    const reader = new Reader(rb, sigMgr2, 1000);

    const largePayload = Buffer.from(Array.from({ length: 250 }, (_, i) => i % 256));

    const readResult: Buffer[] = [];
    let readErr: any = null;

    const readPromise = (async () => {
        try {
            const res = await reader.readFull(250);
            readResult.push(res);
        } catch (e) {
            readErr = e;
        }
    })();

    try {
        const written = await writer.write(largePayload);
        assert.strictEqual(written, 250);
        await readPromise;
    } finally {
        writer.close();
        reader.close();
        try { fs.unlinkSync(filePath); } catch (e) {}
    }

    if (readErr) throw readErr;
    assert.deepStrictEqual(readResult[0], largePayload);
});

File: ringbuf/ringbuf_test.go
package ringbuf

import (
	"bytes"
	"net"
	"os"
	"path/filepath"
	"strings"
	"testing"
	"time"
)

// localSocketPair creates a connected pair of local TCP sockets
// to act as high-fidelity simulated signal connections for unit tests.
func localSocketPair(t *testing.T) (net.Conn, net.Conn) {
	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("failed to listen: %v", err)
	}
	defer ln.Close()

	var c1 net.Conn
	var err1 error
	done := make(chan struct{})
	go func() {
		c1, err1 = ln.Accept()
		close(done)
	}()

	c2, err := net.Dial("tcp", ln.Addr().String())
	if err != nil {
		t.Fatalf("failed to dial: %v", err)
	}
	<-done
	if err1 != nil {
		c2.Close()
		t.Fatalf("failed to accept: %v", err1)
	}
	return c1, c2
}

func TestRingBuffer_Basic(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "shm_test_*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	filePath := filepath.Join(tmpDir, "shm_file")
	rb, err := newRingBuffer(filePath, 2, 64)
	if err != nil {
		t.Fatalf("failed to create ringBuffer: %v", err)
	}
	defer rb.Close()

	rb.Clear()

	c1, c2 := localSocketPair(t)
	defer c1.Close()
	defer c2.Close()

	// Write first slot
	payload1 := []byte("hello world")
	if err := rb.Write(payload1, 42, 10*time.Millisecond, c1); err != nil {
		t.Fatalf("failed to write: %v", err)
	}

	// Read first slot
	readPayload1, seq1, err := rb.Read(10*time.Millisecond, c2)
	if err != nil {
		t.Fatalf("failed to read: %v", err)
	}

	if seq1 != 42 {
		t.Errorf("expected seq 42, got %d", seq1)
	}
	if !bytes.Equal(readPayload1, payload1) {
		t.Errorf("expected %s, got %s", payload1, readPayload1)
	}
}

func TestRingBuffer_TimeoutAndDrop(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "shm_test_*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	filePath := filepath.Join(tmpDir, "shm_file")
	// Buffer has 2 slots, size 64 each
	rb, err := newRingBuffer(filePath, 2, 64)
	if err != nil {
		t.Fatalf("failed to create ringBuffer: %v", err)
	}
	defer rb.Close()

	rb.Clear()

	c1, c2 := localSocketPair(t)
	defer c1.Close()
	defer c2.Close()

	// Write 1st slot (index 0)
	if err := rb.Write([]byte("msg1"), 1, 10*time.Millisecond, c1); err != nil {
		t.Fatalf("failed to write msg1: %v", err)
	}

	// Write 2nd slot (index 1)
	if err := rb.Write([]byte("msg2"), 2, 10*time.Millisecond, c1); err != nil {
		t.Fatalf("failed to write msg2: %v", err)
	}

	// Buffer is now full.
	// Next write should wait on slot 0 and timeout because slot 0 is not read yet.
	err = rb.Write([]byte("msg3"), 3, 20*time.Millisecond, c1)
	if err == nil {
		t.Fatal("expected write timeout/drop error, got nil")
	}

	if !strings.Contains(err.Error(), "write timeout") {
		t.Errorf("unexpected error message: %v", err)
	}

	// The index should have advanced to slot 1 due to the drop/skip
	if rb.GetWriteIndex() != 1 {
		t.Errorf("expected write index to advance to 1, got %d", rb.GetWriteIndex())
	}
}

func TestConnection_Basic(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "shm_test_*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	basePath := filepath.Join(tmpDir, "shm_comm")

	type connResult struct {
		writer *Writer
		reader *Reader
		err    error
	}

	// Create Host connection asynchronously
	hostChan := make(chan connResult, 1)
	go func() {
		w, r, err := NewConnection(basePath, RoleHost, 2, 64, 500*time.Millisecond, 500*time.Millisecond)
		hostChan <- connResult{w, r, err}
	}()

	// Create Plugin connection synchronously
	pluginWriter, pluginReader, err := NewConnection(basePath, RolePlugin, 2, 64, 500*time.Millisecond, 500*time.Millisecond)
	if err != nil {
		t.Fatalf("failed to create Plugin connection: %v", err)
	}
	defer pluginWriter.Close()
	defer pluginReader.Close()

	res := <-hostChan
	if res.err != nil {
		t.Fatalf("failed to create Host connection: %v", res.err)
	}
	hostWriter := res.writer
	hostReader := res.reader
	defer hostWriter.Close()
	defer hostReader.Close()

	// Host writes, Plugin reads
	msgFromHost := []byte("hello from host")
	_, err = hostWriter.Write(msgFromHost)
	if err != nil {
		t.Fatalf("failed to write from host: %v", err)
	}

	readBufPlugin := make([]byte, 100)
	rn, err := pluginReader.Read(readBufPlugin)
	if err != nil {
		t.Fatalf("plugin failed to read: %v", err)
	}
	if !bytes.Equal(readBufPlugin[:rn], msgFromHost) {
		t.Errorf("expected %s, got %s", msgFromHost, readBufPlugin[:rn])
	}

	// Plugin writes, Host reads
	msgFromPlugin := []byte("hello from plugin")
	_, err = pluginWriter.Write(msgFromPlugin)
	if err != nil {
		t.Fatalf("failed to write from plugin: %v", err)
	}

	readBufHost := make([]byte, 100)
	rn, err = hostReader.Read(readBufHost)
	if err != nil {
		t.Fatalf("host failed to read: %v", err)
	}
	if !bytes.Equal(readBufHost[:rn], msgFromPlugin) {
		t.Errorf("expected %s, got %s", msgFromPlugin, readBufHost[:rn])
	}
}

func TestConnection_MultiSlot(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "shm_test_*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	basePath := filepath.Join(tmpDir, "shm_comm")

	type connResult struct {
		writer *Writer
		reader *Reader
		err    error
	}

	// Host connection (4 slots, 64 bytes payload each) asynchronously
	hostChan := make(chan connResult, 1)
	go func() {
		w, r, err := NewConnection(basePath, RoleHost, 4, 64, 500*time.Millisecond, 500*time.Millisecond)
		hostChan <- connResult{w, r, err}
	}()

	// Plugin connection synchronously
	pluginWriter, pluginReader, err := NewConnection(basePath, RolePlugin, 4, 64, 500*time.Millisecond, 500*time.Millisecond)
	if err != nil {
		t.Fatalf("failed to create Plugin connection: %v", err)
	}
	defer pluginWriter.Close()
	defer pluginReader.Close()

	res := <-hostChan
	if res.err != nil {
		t.Fatalf("failed to create Host connection: %v", res.err)
	}
	hostWriter := res.writer
	hostReader := res.reader
	defer hostWriter.Close()
	defer hostReader.Close()

	// Host writes 150 bytes, spanning 3 slots
	payload := make([]byte, 150)
	for i := range payload {
		payload[i] = byte(i)
	}

	n, err := hostWriter.Write(payload)
	if err != nil {
		t.Fatalf("failed to write multi-slot payload: %v", err)
	}
	if n != len(payload) {
		t.Errorf("expected %d bytes written, got %d", len(payload), n)
	}

	// Plugin reads it back fully
	readBuf := make([]byte, 150)
	_, err = pluginReader.ReadFull(readBuf)
	if err != nil {
		t.Fatalf("failed to read multi-slot payload: %v", err)
	}

	if !bytes.Equal(readBuf, payload) {
		t.Error("read payload does not match written payload")
	}
}

func BenchmarkConnection_WriteRead(b *testing.B) {
	tmpDir, err := os.MkdirTemp("", "shm_bench_*")
	if err != nil {
		b.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tmpDir)

	basePath := filepath.Join(tmpDir, "shm_bench")

	type connResult struct {
		writer *Writer
		reader *Reader
		err    error
	}

	hostChan := make(chan connResult, 1)
	go func() {
		w, r, err := NewConnection(basePath, RoleHost, 8, 4096, 500*time.Millisecond, 500*time.Millisecond)
		hostChan <- connResult{w, r, err}
	}()

	pluginWriter, pluginReader, err := NewConnection(basePath, RolePlugin, 8, 4096, 500*time.Millisecond, 500*time.Millisecond)
	if err != nil {
		b.Fatalf("failed to create Plugin connection: %v", err)
	}
	defer pluginWriter.Close()
	defer pluginReader.Close()

	res := <-hostChan
	if res.err != nil {
		b.Fatalf("failed to create Host connection: %v", res.err)
	}
	hostWriter := res.writer
	hostReader := res.reader
	defer hostWriter.Close()
	defer hostReader.Close()

	payload := make([]byte, 1024) // 1KB message
	readBuf := make([]byte, 1024)

	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_, err = hostWriter.Write(payload)
		if err != nil {
			b.Fatalf("write failed: %v", err)
		}
		_, err = pluginReader.ReadFull(readBuf)
		if err != nil {
			b.Fatalf("read failed: %v", err)
		}
	}
}
